src/backend/joanie/core/factories.py

```diff
@@ -288,6 +312,57 @@ def enrollment_end(self):
             )
         ).replace(tzinfo=timezone.utc)
 
+    @staticmethod
+    def get_dates_from_state(state, ref_date=django_timezone.now()):
```

Comment:
I would find it more elegant to take into account ref_date and state in each of the 4 field methods (start, end, enrollment_start and enrollment_end) to generate a date.
For example in start: if state open, start should be ref_date - 1h, if future start should be ref_date + 1h.... and keep the random factor as we have it now?
I see several advantages over what you proposed:
- easier to understand what each date will receive
- does not break forcing fields via attributes
- what you propose calculates all 4 dates and takes start, then calculates all 4 dates again and takes end, etc. Not very efficient and confusing?